File: python/2d.py
import math

class Pos:
	def __init__(self, x, y):
		self.x = x
		self.y = y

	def __repr__(self):
		return f'Pos(x: {self.x}, y: {self.y})'

class Tangent:
	def __init__(self, delta):
		self.delta = delta

	def __repr__(self):
		return f'tan(rad: {self.radiance}, deg: {self.angle})'

	@property
	def radiance(self):
		return math.atan2(self.delta.y, self.delta.x)

	@property
	def angle(self):
		return (self.radiance * (180/math.pi) + 360) % 360

class Delta:
	def __init__(self, source, destination):
		self.source = source
		self.destination = destination

	def __repr__(self):
		return f'Δ(x: {self.x}, y: {self.y})'

	@property
	def x(self):
		return self.destination.x - self.source.x

	@property
	def y(self):
		return self.destination.y - self.source.y

	@property
	def distance(self):
		pass

class Vector:
	def __init__(self, source, angle):
		self.source = source
		self.angle = angle

	def __repr__(self):
		return f'Vector(source: {self.source}, angle: {this.angle})'

	def move(self, distance):
		x_multiplier = math.cos(((self.angle)/180)*math.pi)
		y_multiplier = math.sin(((self.angle)/180)*math.pi)

		return Pos(x=self.source.x + (x_multiplier * distance),
					y=self.source.y + (y_multiplier * distance))

class Distance(Delta):
	def __repr__(self):
		return f'Distance({self.pixels})'

	@property
	def pixels(self):
		return pow(pow(self.destination.x - self.source.x, 2) + pow(self.destination.y - self.source.y, 2), 0.5)

if __name__ == '__main__':
	start = Pos(50, 70)
	destination = Pos(120, 120)

	delta = Delta(start, destination)
	tangent = Tangent(delta)

	new_position = Vector(start, tangent.angle).move(distance=5)
	distance = Distance(start, new_position)

	print('Start: ' + str(start))
	print('Destination: ' + str(destination))
	print('Delta: ' + str(delta))
	print('Tangent: ' + str(tangent))
	print('Degrees: ' + str(int(tangent.angle)))
	print('New pos: ' + str(new_position))
	print('Distance in pixels: ' + str(distance))

File: JavaScript/2d.js
class _Pos {
	constructor(x, y) {
		this.x = x;
		this.y = y;
	}

	__repr__() {
		return `Pos(x: ${this.x}, y: ${this.y})`
	}
}

class _Tangent {
	constructor(delta) {
		this.delta = delta
	}

	__repr__() {
		return `tan(rad: ${this.radiance}, deg: ${this.angle})`
	}

	get radiance() {
		return Math.atan2(this.delta.y, this.delta.x);
	}

	get angle() {
		return (this.radiance * (180 / Math.PI) + 360) % 360
	}

	get degrees() {
		return this.angle()
	}
}

class _Delta {
	constructor(source, destination) {
		this.source = source
		this.destination = destination
	}

	__repr__() {
		return `Δ(x: ${this.x}, y: ${this.y})`
	}

	get x() {
		return this.destination.x - this.source.x;
	}

	get y() {
		return this.destination.y - this.source.y;
	}

	get distance() {
		return;
	}
}

class _Vector {
	constructor(source, angle) {
		this.source = source
		this.angle = angle
	}

	__repr__() {
		let src = str(this.source)
		return `Vector(source: ${src}, angle: ${this.angle})`
	}

	move(distance) {
		let x_multiplier = Math.cos(((this.angle)/180)*Math.PI)
		let y_multiplier = Math.sin(((this.angle)/180)*Math.PI)

		return new _Pos(this.source.x + (x_multiplier * distance),
						this.source.y + (y_multiplier * distance))
	}
}

class _Distance extends _Delta {
	constructor(source, destination) {
		super(source, destination)
	}

	__repr__() {
		return `Distance(${this.pixels})`
	}

	get pixels() {
		return Math.pow(Math.pow(this.destination.x - this.source.x, 2) + Math.pow(this.destination.y - this.source.y, 2), 0.5)
	}
}

function Pos(x, y) {
	return new _Pos(x, y)
}
function Tangent(delta) {
	return new _Tangent(delta)
}
function Delta(source, destination) {
	return new _Delta(source, destination)
}
function Vector(source, angle) {
	return new _Vector(source, angle)
}
function Distance(source, destination) {
	return new _Distance(source, destination)
}
function str(obj) {
	if (typeof obj.__repr__ !== 'undefined')
		return obj.__repr__()
	return obj
}

window.space = {
	'Pos' : Pos,
	'Tangent' : Tangent,
	'Delta' : Delta,
	'Vector' : Vector,
	'Distance' : Distance,
	'str' : str
}
